### PR TITLE
fix: redact private key in broadcast functions and env values

### LINKED ISSUE #3333: script leaks env priv key to terminal w high verbosity
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (3a462eb 2022-09-22T03:37:08.364254Z)

### What command(s) is the bug in?

forge script

### Operating System

macOS (Intel)

### Describe the bug

not sure if this qualifies as a "bug" but I think it's worth discussing. Example cmd/output below for reference

```
λ forge script script/Script.s.sol:MyScript --rpc-url $RPC_URL -vvvv
[⠰] Compiling...
No files changed, compilation skipped
Traces:
  [x] MyScript::run()
    ├─ [0] VM::envUint(PRIVATE_KEY)
    │   └─ ← {{ leaked }}
    ├─ [0] VM::startBroadcast({{ leaked }})
    │   └─ ← ()
    ├─ [x] → new MyFactory@address
    │   ├─ [x] → new MyImpl@address
    │   │   └─ ← x bytes of code
    │   └─ ← x bytes of code
    ├─ [0] VM::stopBroadcast()
    │   └─ ← ()
    └─ ← ()


Script ran successfully.
==========================
Simulated On-chain Traces:

  [x] Create2Deployer::create2()
    ├─ [x] → new MyFactory@address
    │   ├─ [x] → new MyImpl@address
    │   │   └─ ← x bytes of code
    │   └─ ← x bytes of code
    └─ ← address
```

ideally there would be a way to get the trace w/o logging `privKey` to term? but maybe .. maybe these deploy scripts just shouldn't be run w `-vvvv`

(I've always thought having sensitive information in the terminal was less secure than inside a e.g. `.env` but as I'm thinking about it now if someone has access to your terminal screen / output they probably have access to your `.env`? arguably the main downside is the logs might be stored for a long time and searched later, whereas presumably you don't keep your `privKey` stored for a long period of time, only when in use)

### PR DESCRIPTION
**Pull request by mattsse**
## Motivation
Closes #3333

Redact input for cheatcodes that accept a private key:

```console
Traces:
  [27091] CounterScript::run() 
    ├─ [0] VM::startBroadcast(<pk>) 
    │   └─ ← ()
    └─ ← ()

```

### CODE REVIEW

File: evm/src/trace/utils.rs [hunk 1]
```diff
@@ -34,6 +34,16 @@ pub(crate) fn decode_cheatcode_inputs(
         "expectRevert" => {
             decode::decode_revert(data, Some(errors), None).ok().map(|decoded| vec![decoded])
         }
+        "startBroadcast" | "broadcast" => {
```
[the remark this answers is not in the document]
Comment (mds1) [reply evm/src/trace/utils.rs:37]: Maybe we also hide all env cheats and if you need them logged you use console.log?

[the remark this answers is not in the document]
Comment (mattsse) [reply evm/src/trace/utils.rs]: ah right, forgot about response of env codes, let me check where these are handled

**Comment by mattsse** [comment]
also redacted env return values:

```
Traces:
  [28128] CounterScript::run() 
    ├─ [0] VM::envString(PRIV_KEY) 
    │   └─ ← <env var value>
    ├─ [0] VM::startBroadcast(<pk>) 
    │   └─ ← ()
    └─ ← ()

```

File: evm/src/trace/utils.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (mds1) [reply evm/src/trace/utils.rs]: just took a look through forge-std's `Vm.sol`, looks like there's a few others we need, here's a full list of cheats that have a private key input or output (ignoring all the env cheats which it looks like you already got)

```solidity
// Inputs
function sign(uint256,bytes32) external returns (uint8,bytes32,bytes32);
function addr(uint256) external returns (address);
function broadcast(uint256) external;
function startBroadcast(uint256) external;
// Outputs
function deriveKey(string calldata, uint32) external returns (uint256);
function deriveKey(string calldata, string calldata, uint32) external returns (uint256);
function rememberKey(uint256) external returns (address);
```

[the remark this answers is not in the document]
Comment (mattsse) [reply evm/src/trace/utils.rs]: all covered now

File: evm/src/trace/utils.rs [hunk 2]
```diff
@@ -34,6 +36,39 @@ pub(crate) fn decode_cheatcode_inputs(
...
+
+/// Custom decoding of cheatcode return values
+pub(crate) fn decode_cheatcode_outputs(func: &Function, _data: &[u8]) -> Option<String> {
+    if func.name.starts_with("env") {
+        // redacts the value stored in the env var
+        return Some("<env var value>".to_string())
+    }
+    if func.name == "deriveKey" {
+        // redacts derived private key
+        return Some("<pl>".to_string())
```
Comment (mds1) [inline evm/src/trace/utils.rs:71]:
```suggestion
        return Some("<pk>".to_string())
```

**Pull request merged**